### PR TITLE
test_conversors now tests end-to-end decoding

### LINKED ISSUE #52: Bugs in sparse data with missing nominal values
image

Due to the default value str(0), it may causes BadNominalValue() error when data contains some missing nominal values.

### LINKED ISSUE #77: Sparse with quoted values not supported
The ARFF spec includes quoted nominals in their sparse ARFF examples. liac-arff appears to raise an error in this case.

### What I did

```py
import arff
arff.loads('''
@relation foobar

@attribute a real
@attribute b {X,Y,Z}
@attribute c real
@attribute d real
@attribute cls {"class A", "class B"}
@data
{1 X, 3 .5, 4 "class A"}
''')
```

### What I expected

```py
{'data': [[0, 'X', 0, .5, 'class A']], ...}
```

### What I got

```pytb
ValueError                                Traceback (most recent call last)
<ipython-input-2-65109511f99c> in <module>()
      9 @data
     10 {1 X, 3 .5, 4 "class A"}
---> 11 '''
     12 )

/Users/joel/repos/liac-arff/arff.py in loads(s, encode_nominal, return_type)
   1010     decoder = ArffDecoder()
   1011     return decoder.decode(s, encode_nominal=encode_nominal,
-> 1012                           return_type=return_type)
   1013
   1014 def dump(obj, fp):

/Users/joel/repos/liac-arff/arff.py in decode(self, s, encode_nominal, return_type)
    829         try:
    830             return self._decode(s, encode_nominal=encode_nominal,
--> 831                                 matrix_type=return_type)
    832         except ArffException as e:
    833             # print e

/Users/joel/repos/liac-arff/arff.py in _decode(self, s, encode_nominal, matrix_type)
    798             # DATA INSTANCES --------------------------------------------------
    799             elif STATE == _TK_DATA:
--> 800                 data.decode_data(row, self._conversors)
    801             # -----------------------------------------------------------------
    802

/Users/joel/repos/liac-arff/arff.py in decode_data(self, s, conversors)
    350
    351     def decode_data(self, s, conversors):
--> 352         values = self._get_values(s)
    353
    354         if values[0][0].strip(" ") == '{':

/Users/joel/repos/liac-arff/arff.py in _get_values(self, s)
    369         '''(INTERNAL) Split a line into a list of values'''
    370         if _RE_QUOTATION_MARKS.search(s):
--> 371             return _read_csv(s.strip(' '))
    372         else:
    373             return next(csv.reader([s.strip(' ')]))

/Users/joel/repos/liac-arff/arff.py in _read_csv(line)
    575                 raise ValueError(
    576                     'Only whitespace allowed before quoting character at '
--> 577                     'index %d in line: %s' % (i, line))
    578             if quoted is False:
    579                 token = ''

ValueError: Only whitespace allowed before quoting character at index 14 in line: {1 X, 3 .5, 4 "class A"}
```

### Versions

0e17414

### PR DESCRIPTION
**Pull request by jnothman**
Nowhere properly tested decoding values in a sparse context. Focused unit testing of Conversor was the wrong way to go about testing this function, and it allowed us to think end-to-end functionality was working when it wasn't.

As such these tests fail, and I will endeavor to fix the implementations.

So I expect this will:

* fix #77 
* be helpful in completing #76

### CODE REVIEW

**Comment by jnothman** [comment]
These tests now pass.

File: arff.py [hunk 1]
```diff
@@ -367,8 +373,10 @@ def decode_data(self, s, conversors):
 
     def _get_values(self, s):
         '''(INTERNAL) Split a line into a list of values'''
+        if s.rstrip().endswith('}'):
+            return {int(k): v for k, v in _RE_SPARSE_KEY_VALUES.findall(s)}
         if _RE_QUOTATION_MARKS.search(s):
-            return _read_csv(s.strip(' '))
+                return _read_csv(s.strip(' '))
```
Comment (mfeurer) [inline arff.py]: I assume this new indent was unintended?

**Comment by mfeurer** [comment]
I guess this wasn't ready?

**Comment by jnothman** [comment]
Yes, atm it does not fail on bad arff... But you have no test ensuring that such formatting errors are caught. If you provide the test, I'll try provide the fix.

**Comment by mfeurer** [comment]
Okay @jnothman, it took me longer than anticipated (apparently, there were quite a few bugs in the old implementation) but the master branch is now able to handle the examples I gave above and provides unit tests for them.

Detection of wrong quoting is now in the `Conversor` and you might consider updating the code for #76. Also, due to a bugfix with the `Data` object not failing if the data is sparse and has more attributes than declared you need to work around a merge conflict.

Please let me know if these changes help you accomplishing better test and faster data loading, I'd be happy to do more changes if this improves the package.

**Comment by jnothman** [comment]
I had only expected you to provide tests, rather than a fix. I have implemented a fix here, which throws errors on invalid layout of sparse records. I had thought I'd posted a note saying so, but apparently not. As such we have duplicated effort.

**Comment by mfeurer** [comment]
> I have implemented a fix here, which throws errors on invalid layout of sparse records. I had thought I'd posted a note saying so, but apparently not. As such we have duplicated effort.

It seems that I was a bit too eager here, I assumed it would be best to start fixing and improving this from a clean slate as I also uploaded two more fixes in the commit before. Please excuse me here, I hope this is not an issue for you. Overall, I would prefer to replace the current code by your code from this PR as the regex will be much easier to maintain in the future (and is probably much faster).

**Comment by jnothman** [comment]
we can similarly replace _read_csv with a regex approach, btw.

**Comment by jnothman** [comment]
I don't find having a plethora of fine-grained syntax error exception types helpful.

There's a reason why Python only has, I think, two distinct syntax error types (SyntaxError, IndentationError): once there is a error in the syntax, it's very hard to definitively describe what went wrong. Presuming the intention of the user (e.g. that they failed to quote a multi-token nominal) is not really helpful.

It's much easier to provide a fast parser if it does not worry about keeping track of the kind of state needed to discriminate between many types of error; while that can be done by re-running a more diagnostic parser in case of error, I don't see why we would want to engineer that.

Can I please propose that we have one or two syntax errors, and then a type casting error, and that's it?

**Comment by mfeurer** [comment]
@jnothman this is superseded by #76, right?

**Comment by jnothman** [comment]
Yes

**Pull request closed without merging**